### ISSUE TITLE
MCP: Require local size parameters in gl_compute attribute

### ISSUE THREAD

**Issue by Hentropy** [EmbarkStudios/rust-gpu#486]
# Current Syntax
Currently, specifying the local size of a compute shader looks like this:
```rust
#[spirv(gl_compute(local_size_x = 3, local_size_y = 54, local_size_z = 23))]
pub fn main_cs() {}
```
If any (or all) of the local_size_* parameters are not specified they default to 1 which is inefficient.  There isn't really a sensible default here, it's very hardware specific.  In HLSL specifying local size is a hard requirement.

# Proposal
Specifying the dimensions should be a requirement, and the syntax should be cleaned up to improve usability:
```rust
// LocalSize 32, 23, 67
#[spirv(gl_compute(32, 23, 67))]
pub fn main_cs() {}

// Error!
#[spirv(gl_compute)]
pub fn main_cs() {}
```

Unspecified Y or Z dimensions could be assumed to be 1 for the common case of one or two dimensioned computations:
```rust
// LocalSize 32, 1, 1
#[spirv(gl_compute(32))]
pub fn main_cs() {}

// LocalSize 1, 45, 1
#[spirv(gl_compute(1, 45))]
pub fn main_cs() {}
```

**Comment by XAMPPRocky** [comment]
> Specifying the dimensions should be a requirement, and the syntax should be cleaned up to improve usability

Thank you for your issue! I agree that it would be better for this to be hard requirement that causes a compile error, however I don't really agree the that the syntax proposed is "cleaner", or improves usability to me, because I think those numbers look very magic. There' no real indication as to what these numbers should be doing or what they represent.

If I was to pick new syntax, I think my preference would be for the following, it's just as clear as the current version, but is a lot shorter.

```rust
#[spirv(gl_compute(local_size(x = 3, y = 54, z = 23)))]
pub fn main_cs() {}
```

**Comment by Hentropy** [comment]
The local_size bit is redundant since there is only one kind of dimensionality, and there isn't anything else the numbers could be referring to.

An alternative:
```rust
// LocalSize 1, 45, 1
#[spirv(gl_compute(y = 45))]
pub fn main_cs() {}
```
This extends well to LocalSizeId as well, if/when support for that is added:
```rust
// LocalSizeId 
#[spirv(gl_compute(y_id = /* syntax tbd */))]
pub fn main_cs() {}
```

The only other ExecutionModes that apply to gl_compute entry points are the SPV_KHR_float_controls ones.

**Comment by XAMPPRocky** [comment]
> The local_size bit is redundant since there is only one kind of dimensionality, and there isn't anything else the numbers could be referring to.

Just because it's redundant for the compiler, doesn't mean it's redundant for the person reading the code. The syntax needs to be obvious and understandable for people at all skill levels. There are going to be people using `rust-gpu` who are very familiar with Rust and new to GPU programming, experienced GPU devs who are learning Rust, and the syntax needs to be obvious and understandable to both groups.

Our current syntax matches the `layout(local_size_x = 3, local_size_y = 54, local_size_z = 23)` in GLSL, so why would someone know that `gl_compute(32, 23, 67)` maps to that `layout` syntax?

**Comment by Jasper-Bekkers** [comment]
> Our current syntax matches the layout(local_size_x = 3, local_size_y = 54, local_size_z = 23) in GLSL, so why would someone know that gl_compute(32, 23, 67) maps to that layout syntax?

In hlsl it's specified through `[numthreads(x, y, z)]` which is quite nice & understandable imho.

**Comment by khyperia** [comment]
more bikeshedding! thing brought up in the meeting:

`#[spirv(gl_compute(local_size = (1, 2, 3)))]`

**Comment by XAMPPRocky** [comment]
Discussed at the meeting, we're generally in favour of making the parameters required, and having a compile error when missing, and in favour of changing the syntax to something shorter, but still want something that's obvious for readers.

---

Looking at the above two comments makes me think we should at least consider the following.

`#[spirv(gl_compute(numthreads = (1, 2, 3)))]`

**Comment by Hentropy** [comment]
The general consensus seems to be `name = (tuple)`.

### Name
**local_size**
- Pro: similar to GLSL and SPIR-V
- Pro: Rusty name
- Con: not clear what it is unless the reader is already familiar with GLSL or SPIR-V

**numthreads**
- Pro: what HLSL does
- Pro: clear what it's referring to
- Con: non-Rusty name/formatting (ie. not snake_case)

**thread_count**
- Pro: clear what it's referring to
- Pro: Rusty name
- Con: not what GLSL or HLSL does

### Tuple
**Elide trailing ones**
Compute shaders are most likely one dimensional and second most likely two dimensional, so having to always specify the trailing ones in (x, 1, 1) or (x, y, 1) is mildly annoying.  Allowing the trailing ones to be elided is a nice lil UX win.

---
I prefer `thread_count` with trailing one elision.
```rust
#[spirv(gl_compute(thread_count = (43)))]
```

**Comment by Hentropy** [comment]
Implementing `name = (tuple)` requires special attribute parsing in the compiler, but the compiler helpers make it easy to implement the following syntax:
```rust
#[spirv(compute(threads(2)))]
#[spirv(compute(threads(2, 34)))]
#[spirv(compute(threads(2, 34, 4)))]
```
This is concise and idiomatic, so I went with that.
Implemented in #495
